docs/function-default-arguments.md

```diff
@@ -0,0 +1,200 @@
+
+For example
+
+```lua
+function foo(a, b = a)
+    print(a, b)
+end
+foo(1)
+```
+
+will output `1 nil` rather than the potentially expected `1 1`. When evaluating the default value for `b`, the parameter `a` has not yet entered scope and so `= a` attempts to reference `a` from higher scopes. This behavior is analogous to the statement `local a, b = 1, a`.
+
+It is noted that this may not be desired behavior in all cases; one can imagine a function such as the below where it may be beneficial to be able to access the arguments during defaulting:
+
+```lua
+function table.remove(t: Array, pos: number = #t)
+```
+
+This RFC does not currently consider this, though future works may re-investigate this behavior.
+
+### Type Semantics: Type annotations take precedence over default types when performing inference
+When inferring types, an explicit annotation has precedence over the inferred type of the default argument. For example, in the following function the error should be that a `Type 'string' could not be converted into 'number'`.
+
+```lua
+function demo_2(a: number = "demo") end
+```
+
+### Type Semantics: Inferred types from default values take precedence over inferred types from the function body
+If a type has been inferred from the default value, a conflicting type within the function body will cause an error. The following code is invalid
+
+```lua
+function demo_3(a = "demo")
+    a = 1  -- Invalid
+end
+```
+
+### Type Semantics: Inferred types from default values are not sealed
+When a table type is inferred from a default value, it is not sealed. If the function body makes modifications, these are unified with the inferred type.
+
+In the following example, `point` has a final type of `{x:number,y:number,z:number}`.
+
+```lua
+function demo_unsealed(point = {x = 0, y = 0})
+    point.z = 0  -- Ok
+end
+```
+
+If sealing is desired, the programmer may be explicit about this using either an annotation or a cast. The former takes precedence over the inferred type entirely, and the latter ensures the inferred type is sealed. Rather than a unique semantic this is a consequence of previously defined semantics.
+
+```lua
+type PointType = {
+    x: number,
+    y: number,
+}
+function demo_sealed_1(point: PointType = {x = 0, y = 0})
+    point.z = 0  -- Invalid
+end
+function demo_sealed_2(point = {x = 0, y = 0} :: PointType)
+    point.z = 0  -- Invalid
+end
+```
+
+### Language implementation
+Within the AST, it would likely be simpler to add a new `AstArray<AstExpr*> argsDefaults` to `AstExprFunction` alongside `args`, rather than modify `args` to be an `std::pair` due to the existing widespread usage of `args`.
+
+Rather than implement this as a feature of the `CALL` instruction within Luau's VM it is instead suggested by this RFC to implement this as part of the compiler. This both increases compatibility (as the VM remains unchanged) and makes it easier to allow *any* expression to be used.
```

Comment:
So does this mean `function f(x: number = 3)` means that I call `f()` and it compiles as `f(3)`?

What happens then if I use that function as a first class value?

What happens then in this case?
```lua
local function g(callback: (number) -> ())
    callback()
end

g(f)
```

Reply:
I ask because surely something with default parameters cannot have a non-optional argument as part of its type and *also* be a first class value?

Reply:
The suggested implementation essentially injects a series of `if x == nil then x = default end` statements at the start of the function body, so passing `f` around as a first class value is totally fine. If we run
```lua
-- !strict
function f(x: number = 3)
    print(x)
end

local function g(callback: (number) -> ())
    callback()
end

g(f)
```
we get `3` on stdout as expected.

We do however get a type error of `TypeError: Argument count mismatch. Function 'callback' expects 1 argument, but none are specified` because the first class value `f` will have a type of `(number?) -> ...any` as defined there (to indicate that we need not provide that argument/could provide `nil`).

Amending g to `local function g(callback: (number?) -> ())` that type error goes away.

Reply:
I've [added a section](https://github.com/Bottersnike/luau-rfcs/blob/function-default-arguments/docs/function-default-arguments.md#type-semantics-arguments-with-a-default-value-are-implicitly-optional) that hopefully formalises that brief explanation into the RFC. lmk if there's still confusion (or another type interaction I've not formalised!).